lib/Conversion/TTIRToTTMetal/TTIRToTTMetal.cpp

```diff
@@ -1344,6 +1370,268 @@ class TTIRToTTMetalDispatchRewriter : public OpRewritePattern<ttir::GenericOp> {
+                            op.getInputs().size())
+                            .value();
+    auto inputCB =
+        mlir::cast<ttkernel::CBType>(*rewrittenBlockArgumentTypes.begin());
+    auto inputTT = mlir::cast<TileType>(inputCB.getMemref().getElementType());
+    assert(inputTT);
+
+    // Single tile memref that will be used for the scaler.
+    MemRefType tileMemref = MemRefType::get(
+        {1, 1}, inputTT,
+        mlir::AffineMap::getMultiDimIdentityMap(2, op->getContext()),
+        MemorySpaceAttr::get(op->getContext(), MemorySpace::DeviceL1));
+
+    auto scalerCBType = ttkernel::CBType::get(
+        op.getContext(), scalerCBPort,
+        op.getSystemDesc().getChipDescs().front().getGlobalL1RegionAddress(),
+        tileMemref, true /* is_internal */);
+    auto scalerCB = dmBlock->addArgument(scalerCBType, op.getLoc());
+
+    auto reduceKind = kernelOp.getKind();
+    float scalerValue = 0.;
+    if (reduceKind == "sum" || reduceKind == "max") {
+      scalerValue = 1.;
+    }
+
+    if (reduceKind == "mean") {
+      int64_t numElements = 1;
+      auto inputType =
+          mlir::cast<RankedTensorType>(op.getInputs()[0].getType());
+
+      for (int64_t dim = 0; dim < inputType.getRank(); ++dim) {
+        auto iteratorType =
+            mlir::cast<IteratorTypeAttr>(op.getIteratorTypes()[dim]);
+        if (iteratorType.getValue() == IteratorType::Reduction) {
+          numElements *= inputType.getShape()[dim];
+        }
+      }
+      scalerValue = 1. / numElements;
+    }
+
+    generateScaler(op.getLoc(), dmBlock, scalerCB, scalerValue);
+
+    return scalerCBType;
+  }
+
+  // Given float value fValue, generate a scaler tile with value fValue. Tile
+  // should be populated in pattern such that only first row of each face is
+  // populated with fValue, while the rest of the tile is zeroed out.
+  static void generateScaler(Location loc, Block *block,
+                             BlockArgument &scalerCB, float fValue) {
+    OpBuilder builder(block, block->begin());
+
+    // Assumption is that scalerCB is tile of F16/BF16 values. Converting from
+    // float to 2byte float is truncating mantissa bits. To support lesser
+    // fortmats we need to add conversion from float to those formats, which is
+    // trickier than just truncating mantissa bits.
+    uint32_t iValue = *reinterpret_cast<uint32_t *>(&fValue);
+    iValue >>= 16;
+    uint16_t iValue16 = iValue & 0xFFFF;
+    auto scalerConst = builder.create<arith::ConstantOp>(
+        loc, builder.getI32Type(),
+        builder.getI32IntegerAttr(iValue16 | (iValue16 << 16)));
+
+    // Reserve single tile.
+    auto oneConst = builder.create<arith::ConstantOp>(
+        loc, builder.getI32Type(), builder.getI32IntegerAttr(1));
+    builder.create<ttkernel::CBReserveBackOp>(loc, scalerCB, oneConst);
+
+    // Prepare zero region read.
+    auto zerosBase = builder.create<ttkernel::MacroOp>(
+        loc, builder.getI32Type(), "MEM_ZEROS_BASE");
```

Comment:
I think it might be better to pretend this is a real function `memZerosBase` and not have a concept of a `MacroOp` at all.

Reply:
Not sure what do you mean exactly? 
Function emits call like `memZerosBase()` while we need only macro name `MEM_ZEROS_BASE`. Or do you suggest to wrap this in a `func.func`?

Reply:
I just mean from the dialect perspective, we should have `MemZerosBaseOp` which happens to be emitted in emitc as all caps.

Reply:
I can do that, but I am wondering what do we gain from that, isn't it better to have a single MacroOp that can be used for any C++ macro we have in kernels?